hax/hax/util.py

```diff
@@ -843,6 +843,102 @@ def get_node_encl_fid(self, node: str) -> Optional[Fid]:
+            LOG.debug('Setting node=%s in KV with state=%s',
+                      node_fid, value['state'])
+            self.kv.kv_put(node['Key'], json.dumps(value))
+
+    @repeat_if_fails()
+    def set_encl_state(self, encl_fid: Fid, status: ServiceHealth) -> None:
+        # Example,
+        # {
+        #    "key": "m0conf/sites/0x5300000000000001:0x1/
+        #            racks/0x6100000000000001:0x2/encls/
+        #            0x6500000000000001:0x4",
+        #    "value": "{\"node\": \"0x6e00000000000001:0x3\",
+        #               \"state\": \"M0_NC_UNKNOWN\"}"
+        # }
+        node_items = self.get_all_sites()
+        regex = re.compile(
+            f'^m0conf\\/.*\\/{encl_fid}$')
+        for encl in node_items:
+            match_result = re.match(regex, encl['Key'])
+            if not match_result:
+                continue
+            value = json.loads(encl['Value'])
+            value['state'] = self.get_device_state(status)
+            LOG.debug('Setting enclosure=%s in KV with state=%s',
+                      encl_fid, value['state'])
+            self.kv.kv_put(encl['Key'], json.dumps(value))
+
+    @repeat_if_fails()
+    def set_ctrl_state(self, ctrl_fid: Fid, status: ServiceHealth) -> None:
+        # Example,
+        # {
+        #    "key": "m0conf/sites/0x5300000000000001:0x1/
+        #            racks/0x6100000000000001:0x2/encls/
+        #            0x6500000000000001:0x4/ctrls/
+        #            0x6300000000000001:0x5",
+        #    "value": "{\"state\": \"M0_NC_UNKNOWN\"}"
+        # }
+        ctrl_items = self.get_all_sites()
+        regex = re.compile(f'^m0conf\\/.*\\/ctrls/{ctrl_fid}$')
+        for ctrl in ctrl_items:
+            match_result = re.match(regex, ctrl['Key'])
+            if not match_result:
+                continue
+            value = json.loads(ctrl['Value'])
+            value['state'] = self.get_device_state(status)
+            LOG.debug('Setting ctrl=%s in KV with state=%s',
+                      ctrl_fid, value['state'])
+            self.kv.kv_put(ctrl['Key'], json.dumps(value))
+
+    @repeat_if_fails()
+    def get_ctrl_state(self, ctrl_fid: Fid) -> int:
+        ctrl_to_ha_state_map = {
+            'unknown': HaNoteStruct.M0_NC_UNKNOWN,
+            'online': HaNoteStruct.M0_NC_ONLINE,
+            'offline': HaNoteStruct.M0_NC_TRANSIENT,
+            'failed': HaNoteStruct.M0_NC_FAILED}
+
+        ctrl_items = self.get_all_sites()
+        regex = re.compile(f'^m0conf\\/.*\\/ctrls/{ctrl_fid}$')
+        for ctrl in ctrl_items:
+            match_result = re.match(regex, ctrl['Key'])
+            if not match_result:
+                continue
+            val = json.loads(ctrl['Value'])
+            LOG.debug('Controller=%s state=%s', ctrl_fid, val['state'])
+            if str(val['state']).lower() in ('unknown', 'offline', 'failed'):
+                return ctrl_to_ha_state_map[str(val['state']).lower()]
```

Comment:
`str(val['state']).lower()` is evaluated multiple times. If you extract it as a variable, that would increase readability.